### PR TITLE
✨ Add image database upload 

### PR DESCRIPTION
**Pull request by gabriel-dantas98**
## Context ☕ 

One of the most important things when creating questions within the Q&A is being able to post screenshots of error pages or even architectural drawings, so we would like to add this functionality to the plugin.

## How

- [x] Addition of 2 new endpoints for sending and receiving files.
- [x] Creating a new table.
- [x] Update app-config parameters

## Tests

image
image
image
image

Check Demo UPDATED 05/04 [[LINK](https://drive.google.com/file/d/1UlsN96n-H3tAIWhTWOrIqcWdca0leYZw/preview)]

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -32,25 +32,21 @@
     ]
   },
   "devDependencies": {
-    "@backstage/cli": "^0.22.6",
+    "@backstage/cli": "^0.22.5",
```
Comment (drodil) [inline package.json:35]: Please sync your fork, the plugin was already following the backstage 1.13.0

File: plugins/qeta-backend/src/service/router.ts [hunk 2]
```diff
@@ -171,10 +175,6 @@ export async function createRouter({
     });
   };
 
-  const permissionIntegrationRouter = createPermissionIntegrationRouter({
```
Comment (drodil) [inline plugins/qeta-backend/src/service/router.ts:174]: Why is this removed?

File: plugins/qeta-backend/src/database/QetaStore.ts [hunk 3]
```diff
@@ -91,6 +91,15 @@ export interface TagResponse {
   tag: string;
   questionsCount: number;
 }
+export interface Attachment {
+  id: number;
+  locationType: string;
+  locationUri: string;
+  binaryImage: Buffer;
```
Comment (drodil) [inline plugins/qeta-backend/src/database/QetaStore.ts:98]: I think this also requires the mimeType of the file to be able to show those properly later

File: plugins/qeta-backend/src/service/router.ts [hunk 4]
```diff
@@ -744,6 +742,52 @@ export async function createRouter({
...
+    const { id } = request.params;
+
+    const attachment = await database.getAttachment(Number(id));
+    const imageBuffer = attachment[0].binaryImage;
+
+    const imageName = `${storageImagesFolderType}/image-${id}-${Date.now()}`;
+
+    fs.writeFileSync(imageName, imageBuffer);
+
+    return response.sendFile(imageName);
```
Comment (drodil) [inline plugins/qeta-backend/src/service/router.ts:788]: Why not just return the image as attachment directly instead first writing it to fs? So something like:
```ts
res.writeHead(200, {'Content-Type': attachment.mimeType, 'Content-Length': imageBuffer.length});
res.end(imageBuffer);
```

File: plugins/qeta-backend/src/service/router.ts [hunk 5]
```diff
@@ -744,6 +742,52 @@ export async function createRouter({
...
+  // GET /attachments/:id
+  router.get('/attachments/:id', async (request, response) => {
+    const storageImagesFolderType =
+      config.getOptionalString('qeta.storage.folder') ||
+      '/tmp/backstage-qeta-images';
+
+    const { id } = request.params;
+
+    const attachment = await database.getAttachment(Number(id));
+    const imageBuffer = attachment[0].binaryImage;
```
Comment (drodil) [inline plugins/qeta-backend/src/service/router.ts:782]: Selecting the first in array should be moved to database side

File: plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx [hunk 6]
```diff
@@ -1,23 +1,64 @@
...
+  }
+  return response.json();
+};
+
+const imageUpload = (config: Config) => {
+  // eslint-disable-next-line func-names
+  return async function* (data: ArrayBuffer) {
+    const { imageURL } = await submitImage(
+      config,
+      new Blob([data], { type: 'text/plain' }),
```
Comment (drodil) [inline plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx:38]: You need to figure out the proper mime type for this in the upload phase. I think users should also allow selecting which kind of files they allow the users to upload in configuration. Now they can basically upload anything. Also, file size limit would be a nice addition to prevent misuse with some large files.

File: plugins/qeta-backend/configSchema.d.ts [hunk 7]
```diff
@@ -6,5 +6,14 @@ export interface Config {
      * @visibility backend
      */
     allowAnonymous: boolean;
+    /**
+     * Configuration about images attachments storage
+     *
+     * @visibility backend
+     */
+    storage: {
+      type?: string;
```
Comment (drodil) [inline plugins/qeta-backend/configSchema.d.ts:15]: You could have this typed out: 
```ts
type: 'database' | 'filesystem';
```

File: plugins/qeta-backend/migrations/20230425-attachments.js [hunk 8]
```diff
@@ -0,0 +1,20 @@
+/**
+ * @param {import('knex').Knex} knex
+ */
+exports.up = async function up(knex) {
+  await knex.schema.createTable('question_attachments', table => {
+    table.increments('id').primary();
+    table.string('locationType').notNullable();
+    table.string('locationUri').nullable();
```
Comment (drodil) [inline plugins/qeta-backend/migrations/20230425-attachments.js:8]: add `mimeType`

File: plugins/qeta-backend/src/database/DatabaseQetaStore.ts [hunk 9]
```diff
@@ -544,6 +545,30 @@ export class DatabaseQetaStore implements QetaStore {
...
+          created: new Date(),
+        },
+        ['id'],
+      )
+      .into('question_attachments');
+
+    return attachments;
+  }
+
+  async getAttachment(uuid: number): Promise<Attachment[]> {
```
Comment (drodil) [inline plugins/qeta-backend/src/database/DatabaseQetaStore.ts:566]: `uuid` is not number, change to `id`

File: plugins/qeta-backend/src/database/DatabaseQetaStore.ts [hunk 10]
```diff
@@ -544,6 +545,30 @@ export class DatabaseQetaStore implements QetaStore {
...
 
+  async postAttachment(
+    locationType: string,
+    binaryImage: Buffer,
+  ): Promise<Attachment[]> {
+    const attachments: Attachment[] = await this.db
+      .insert(
+        {
+          locationType,
+          binaryImage,
```
Comment (drodil) [inline plugins/qeta-backend/src/database/DatabaseQetaStore.ts:556]: This should only set binaryImage to database if the config `storage.type` is `database`. We don't want to save the file both in the filesystem and in database.

File: plugins/qeta-backend/src/database/QetaStore.ts [hunk 11]
```diff
@@ -321,4 +330,11 @@ export interface QetaStore {
    * Returns all used tags for questions
    */
   getTags(): Promise<TagResponse[]>;
+
+  postAttachment(
+    locationType: string,
+    binaryImage: Buffer,
+  ): Promise<Attachment[]>;
+
+  getAttachment(uuid: number): Promise<Attachment[]>;
```
Comment (drodil) [inline plugins/qeta-backend/src/database/QetaStore.ts:339]: Change to `Promise<Attachment>` instead

File: plugins/qeta-backend/src/database/QetaStore.ts [hunk 12]
```diff
@@ -321,4 +330,11 @@ export interface QetaStore {
    * Returns all used tags for questions
    */
   getTags(): Promise<TagResponse[]>;
+
+  postAttachment(
+    locationType: string,
+    binaryImage: Buffer,
+  ): Promise<Attachment[]>;
```
Comment (drodil) [inline plugins/qeta-backend/src/database/QetaStore.ts:337]: Change to `Promise<Attachment>` instead

File: plugins/qeta-backend/src/service/router.ts [hunk 13]
```diff
@@ -744,6 +742,52 @@ export async function createRouter({
     response.send(tags);
   });
 
+  // POST /attachments
+  router.post(
+    '/attachments',
+    multerUploaderMiddleware(config, 'image'),
+    async (request, response) => {
+      const backendBaseUrl = config.getString('backend.baseUrl');
+      const qetaUrl = `${backendBaseUrl}/api/qeta/attachments`;
```
Comment (drodil) [inline plugins/qeta-backend/src/service/router.ts:751]: Checks needed here for the image size and mime type.

File: plugins/qeta-backend/src/service/router.ts [hunk 14]
```diff
@@ -744,6 +742,52 @@ export async function createRouter({
...
+
+      const storageType =
+        config.getOptionalString('qeta.storage.type') || 'database';
+
+      const imageBuffer = await fs.promises.readFile(`${request.file?.path}`);
```
Comment (drodil) [inline plugins/qeta-backend/src/service/router.ts:756]: I don't really follow what is this reading? The file is already included in the request so what is read from the filesystem?

**Review by drodil — CHANGES_REQUESTED** [review]
Thanks! Nice to see this getting forward! I had a couple of points in the review, could you check those out?

I have a bit of a misunderstanding about how the file upload is working as it seems to write the filesystem and database at the same time while the user should be able to choose which one to use. For the future, this can be for example AWS S3 so this should be somehow abstracted for example behind interface AttachmentUploadHandler.

Also having some sane defaults for allowed mime types (only images) and file size (5MB) + possibility to configure this is crucial. Otherwise, we might end up polluting the database/filesystem with files that take up too many resources or cannot even be used. 

Later on, it would also be nice to have a sort of clean-up for the attachments that are not used by any question/answer. For this, it might be that we need to create join tables between questions/answers and the attachments. Another option is to go through the contents of all the questions and answers and find links to attachments that are still used but it's not as powerful.

**Comment by drodil** [comment]
Oh and checked the demo, the upload bar in the MarkdownEditor needs some theming to make it blend in to the backstage :)

**Comment by gabriel-dantas98** [comment]
Hey @drodil 😄 thanks for the review, I got some ideas from your feedback and I believe we now have a much better version.
I ended up deciding to remove the use of multer as we would have to customize a lot of things for our use case.

Added validations for file type and size.

Now, database and filesystem integration are more clearly separated.

And we have more fields in the database.

I was also able to update the dark mode compatibility.

File: plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx [hunk 15]
```diff
@@ -1,23 +1,104 @@
...
+const imageUpload = (config: Config, erroAlert: ErrorApi) => {
+  const maxSizeImage =
+    config?.getOptionalNumber('qeta.storage.maxSizeImage') ||
+    DEFAULT_IMAGE_SIZE_LIMIT;
+
+  // eslint-disable-next-line func-names
+  return async function* (data: ArrayBuffer) {
+    const fileType = await FileType.fromBuffer(data);
+
+    if (fileType && !SUPPORTED_FILES_TYPES.includes(fileType?.mime)) {
```
Comment (drodil) [inline plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx:60]: I think this check should be in the backend side as the API is open for external calls as well. This means that users could upload images without even asking a question. Not sure if there's any way to enforce this..

File: plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx [hunk 16]
```diff
@@ -1,23 +1,104 @@
...
+      erroAlert.post(new Error(`Image type (${fileType.mime}) not supported.`));
+      return false;
+    }
+
+    if (data.byteLength > maxSizeImage) {
```
Comment (drodil) [inline plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx:65]: Same here, move the check to backend return http code 400

File: plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx [hunk 17]
```diff
@@ -1,23 +1,104 @@
...
+      erroAlert.post(
+        new Error(
+          `Image larger than ${maxSizeImage} bytes try to make it smaller before uploading.`,
+        ),
+      );
+      return false;
+    }
+    const { locationUri } = await submitImage(
+      config,
+      new Blob([data], { type: 'text/plain' }),
```
Comment (drodil) [inline plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx:75]: Why text/plain even you know the type here already?

File: plugins/qeta-backend/migrations/20230425-attachments.js [hunk 18]
```diff
@@ -0,0 +1,24 @@
+/**
+ * @param {import('knex').Knex} knex
+ */
+exports.up = async function up(knex) {
+  await knex.schema.createTable('question_attachments', table => {
```
Comment (drodil) [inline plugins/qeta-backend/migrations/20230425-attachments.js:5]: I would change this to just 'attachments' as the 'question_attachments' means a join table in my opinion. Also users probably should be able to upload images also to answers, right?

File: plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx [hunk 19]
```diff
@@ -1,23 +1,104 @@
...
+export const submitImage = async (
+  config: Config,
+  file: Blob,
+  erroAlert: ErrorApi,
+) => {
+  const backendBaseUrl = config.getString('backend.baseUrl');
+  const qetaUrl = `${backendBaseUrl}/api/qeta/attachments`;
+  const formData = new FormData();
+
+  formData.append('image', file);
```
Comment (drodil) [inline plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx:31]: Maybe add question/answer id to the data? This would allow to create relation in database so that when question or answer is deleted, it would automatically delete the attachments as well if they are stored in database. For attachments in filesystem.. Not sure how to clean those..

File: plugins/qeta-backend/src/service/router.ts [hunk 20]
```diff
@@ -744,6 +755,76 @@ export async function createRouter({
...
+        imageBuffer = attachment.binaryImage;
+      } else {
+        imageBuffer = await fs.promises.readFile(attachment.path);
+      }
+
+      if (!imageBuffer) {
+        response.status(500).send('Image buffer is undefined');
+      }
+
+      response.writeHead(200, {
```
Comment (drodil) [inline plugins/qeta-backend/src/service/router.ts:818]: Awesome, looks much better now!

File: plugins/qeta-backend/src/service/router.ts [hunk 21]
```diff
@@ -744,6 +755,76 @@ export async function createRouter({
...
+      const file: File = {
+        name: fileRequest.fieldName,
+        path: fileRequest.path,
+        buffer: fileBuffer,
+        mimeType: mimeType?.mime.toString() || '',
+        ext: mimeType?.ext.toString() || '',
+        size: fileBuffer.byteLength || 0,
+      };
+
+      if (storageType === 'database') {
```
Comment (drodil) [inline plugins/qeta-backend/src/service/router.ts:790]: This is great way to keep this extendible to other storage methods, good work!

File: plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx [hunk 22]
```diff
@@ -1,23 +1,104 @@
...
 import { MarkdownContent } from '@backstage/core-components';
 import 'react-mde/lib/styles/css/react-mde.css';
 import 'react-mde/lib/styles/css/react-mde-editor.css';
 import 'react-mde/lib/styles/css/react-mde-toolbar.css';
 import { useStyles } from '../../utils/hooks';
+import FileType from 'file-type';
+import { ErrorApi, errorApiRef, useApi } from '@backstage/core-plugin-api';
+
+const DEFAULT_IMAGE_SIZE_LIMIT = 2500000;
+const SUPPORTED_FILES_TYPES = [
```
Comment (drodil) [inline plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx:15]: If possible, allow these to be set in the app configuration

**Review by drodil — CHANGES_REQUESTED** [review]
Almost there! Couple of notes about cleaning the unnecessary attachments and checks. Thank you!

File: plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx (hunk 15 above)
Comment (gabriel-dantas98) [reply plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx:60]: We can move on the backend.

I just kept it in the frontend, so we don't have to make a request that we can already validate before the request.

But it makes sense to only use it on the backend in case the endpoints are used in other use cases. 😉

File: plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx (hunk 17 above)
Comment (gabriel-dantas98) [reply plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx:75]: I probably ended up forgetting after the refactor I'll update 😅

File: plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx (hunk 19 above)
Comment (gabriel-dantas98) [reply plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx:31]: That would be great, but I don't know how I would add it since the upload is done before the question is created.

In the cases of answers and editing, we were probably able to add.

For the case of the filesystem, we saved the path information locally in the database, we can use this information to delete it locally in a specific way.

File: plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx (hunk 22 above)
Comment (gabriel-dantas98) [reply plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx:15]: Image size is already being set the types of images I can add

File: plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx (hunk 19 above)
Comment (drodil) [reply plugins/qeta/src/components/MarkdownEditor/MarkdownEditor.tsx:31]: Oh true. So when posting the question, it should actually post the attachment IDs in that data.

File: plugins/qeta-backend/src/service/router.ts [hunk 23]
```diff
@@ -38,6 +48,14 @@ export interface RouterOptions {
   permissions?: PermissionEvaluator;
 }
 
+const DEFAULT_IMAGE_SIZE_LIMIT = 2500000;
+const SUPPORTED_FILES_TYPES = [
```
Comment (gabriel-dantas98) [inline plugins/qeta-backend/src/service/router.ts:52]: These are just to have a default value in case whoever uses the plugin doesn't fill in the app-config. To ease the use of the plugin for those who are not familiar with it.

**Comment by drodil** [comment]
Thanks for the changes! I will check this once again start of next week with proper time and I think it's ready to be merged.

**Review by drodil — APPROVED** [review]
